File: Paradigmas de linguagem de programação/Lista de Exercicio - PLP/Questão 12/Python/Produto.py
# Q12) Em Python,
# sobrecarregue o operador + para somar dois objetos Produto baseados no preço. Em Java
# e Golang, crie métodos que permitam essa funcionalidade.

class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco

    def __add__(self, other):
        return Produto(self.nome + " & " + other.nome, self.preco + other.preco)

    def __str__(self):
        return f"Produto: {self.nome}, Preço: {self.preco}"

produto1 = Produto("Produto A", 50)
produto2 = Produto("Produto B", 30)

produto_soma = produto1 + produto2 
print(produto_soma)

File: Paradigmas de linguagem de programação/Lista de Exercicio - PLP/Questão 12/Java/Produto.java
// Q12) Em Python,
// sobrecarregue o operador + para somar dois objetos Produto baseados no preço. Em Java
// e Golang, crie métodos que permitam essa funcionalidade.

class Produto {
    private String nome;
    private double preco;

    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public Produto somar(Produto outro) {
        String nomeCombinado = this.nome + " & " + outro.nome;
        double precoCombinado = this.preco + outro.preco;
        return new Produto(nomeCombinado, precoCombinado);
    }

    @Override
    public String toString() {
        return "Produto: " + nome + ", Preço: " + preco;
    }

    public static void main(String[] args) {
        Produto produto1 = new Produto("Produto A", 50);
        Produto produto2 = new Produto("Produto B", 30);

        Produto produtoSoma = produto1.somar(produto2);
        System.out.println(produtoSoma);
    }
}